Solutions/Tanium/Playbooks/Tanium-CloseTHRAlert/azuredeploy.json

```diff
@@ -0,0 +1,131 @@
+              "type": "ApiConnectionWebhook",
+              "inputs": {
+                "body": {
+                  "callback_url": "@{listCallbackUrl()}"
+                },
+                "host": {
+                  "connection": {
+                    "name": "@parameters('$connections')['azuresentinel']['connectionId']"
+                  }
+                },
+                "path": "/incident-creation"
+              }
+            }
+          },
+          "actions": {
+            "Add_comment_to_incident_(V3)": {
+              "runAfter": {
+                "HTTP": [
+                  "Succeeded"
+                ]
+              },
+              "type": "ApiConnection",
+              "inputs": {
+                "body": {
+                  "incidentArmId": "@triggerBody()?['object']?['id']",
+                  "message": "<p>@{body('HTTP')}</p>"
+                },
+                "host": {
+                  "connection": {
+                    "name": "@parameters('$connections')['azuresentinel']['connectionId']"
+                  }
+                },
+                "method": "post",
+                "path": "/Incidents/Comment"
+              }
+            },
+            "Entities_-_Get_Hosts": {
+              "runAfter": {},
+              "type": "ApiConnection",
+              "inputs": {
+                "body": "@triggerBody()?['object']?['properties']?['relatedEntities']",
+                "host": {
+                  "connection": {
+                    "name": "@parameters('$connections')['azuresentinel']['connectionId']"
+                  }
+                },
+                "method": "post",
+                "path": "/entities/host"
+              }
+            },
+            "HTTP": {
+              "runAfter": {
+                "Entities_-_Get_Hosts": [
+                  "Succeeded"
+                ]
+              },
+              "type": "Http",
+              "inputs": {
+                "body": "@triggerBody()",
+                "headers": {
+                  "Authorization": "[parameters('ForwarderAPIToken')]"
+                },
+                "method": "POST",
+                "uri": "[concat('https://', parameters('TaniumServerHost'), '/playbook?type=close_thr_alert')]"
+              }
+            }
+          },
+          "outputs": {}
+        },
+        "parameters": {
+          "$connections": {
+            "value": {
+              "azuresentinel": {
+                "connectionId": "[resourceId('Microsoft.Web/connections', parameters('ConnectionName'))]",
+                "connectionName": "[parameters('ConnectionName')]",
+                "id": "[concat('/subscriptions/', subscription().subscriptionId, '/providers/Microsoft.Web/locations/', resourceGroup().location, '/managedApis/azuresentinel')]"
+              }
+            }
+          }
+        }
+      },
+      "dependsOn": [
+        "[resourceId('Microsoft.Web/connections', parameters('ConnectionName'))]"
```

Comment:
Connection is not present in the ARM Template. Deployment failed and thus couldn't proceed with further testing.
Same is true for other playbooks as well.

Reply:
Here is what we do:
https://github.com/tanium/Azure-Sentinel/blob/tanium-dev/tanium_scripts/fix_maintemp.sh

After the build step we modify the `mainTemplate.json` a bit because there are somethings that we cannot change via the input.json. I could not figure out how to add these by tweaking the input.json...
https://github.com/tanium/Azure-Sentinel/blob/tanium-dev/tanium_scripts/fix_maintemp.sh#L14-L16

Reply:
There are few things we need to do here:
1. Each playbook should have the required connections defined within their respective ARM Templates. There is no need to put them separately in Data folder. (See [CiscoUmbrella](https://github.com/Azure/Azure-Sentinel/blob/master/Solutions/CiscoUmbrella/Package/mainTemplate.json) for an example)
2. Once #1 is completed, you can proceed with running the packing tool using the input.json file.
3. There should not be any **Data** folder as mentioned above with the connections. There is always a **Data Connector** folder with the required data connector. In this case, I can see there is a Logic App based connector, is my understanding correct? If yes, connection(s) are missing from this ARM Template as well. Please add the connection within the workflow ARM Template to avoid any issues during the deployment of the connector. I understand that this would bring in duplicates resources of same type in your mainTemplate.json file but that is totally acceptable.

Reply:
@devikamehra I'm a little confused by points 1 and 2 in your response. It looks like our maintemplate.json has the connections defined. So maybe we should simply delete the `AzureSentinelConnector.json` , `TaniumSentinelConnector.json`, and `APIConnectors.json`. 

Is there anyway we could meet and screen share to discuss further? 


For point 3, we are sending data into Log Analytics from our platform, so we don't have a Data Connector like how other venders do. 

For point 4, That is a logic app, but it's not defined as a playbook within the azure portal because its not triggered from an incident. For example, all of the other playbooks are triggered from an incident as shown here: 
![image](https://user-images.githubusercontent.com/11982273/176948077-81f31038-d8d9-44ae-aae1-aa175940ab30.png)

The file in question (https://github.com/Azure/Azure-Sentinel/blob/5307dff2282d94fe716f1b2e08aafaef38597084/Solutions/Tanium/Data/AddCommentLA.json) is a logic app that has a different trigger as shown here: 
![image](https://user-images.githubusercontent.com/11982273/176948154-0cfbbcc3-1843-46bf-9c5c-f1d4d5b9ae4f.png)

Reply:
@ajcollins0 
With point 1, what I meant is that the deployment of the playbook should be successful. When I deploy any one of the playbooks, example [Tanium-AddComment](https://raw.githubusercontent.com/Azure/Azure-Sentinel/07ac54eda970d0a464c2f0816edc749d08b9a877/Solutions/Tanium/Playbooks/Tanium-AddComment/azuredeploy.json), I get the following error.

![image](https://user-images.githubusercontent.com/10944026/178221322-3838bf93-a257-4411-9330-2e043878694b.png)

This is because, the ARM Template has the dependsOn clause for the resource which does not exist in the ARM template. We have to choose from the following two options:

1. We ask the user to individually deploy this connection before-hand and then install playbook without this dependsOn clause.
2. **This is the recommended option.** Include the connection resource for the deployment within the ARM template itself. Whenever the user deploys this ARM Template, they will get both the logic app and the connection.

Please let me know in case more information is needed here.

In point 2, there should be re-packaging of the solution with the updates made in point 1.

In point 3 and 4, can you please explain how are we getting the data into the desired tables? Also, for the file in question, what is the operation we are applying using this logic app? This would help us define whether we should consider this as a playbook or a data connector or none.

Reply:
@devikamehra  can you send me an example of how I can include the connection resource inside the Arm template? Every playbook should rely on the same API Connection object, so I will need to ensure each playbook has the same reference to this connection. Once I get this information I will include it and repackage the solution. 


`can you please explain how are we getting the data into the desired tables?` 
Our software has a way to ship data from our product into other products. It's called Tanium Connect. The general documentation is found here: https://docs.tanium.com/connect/connect/index.html. We are releasing a Microsoft Log Analytics connector where the customers can input their Azure information and we will send data from our platform directly into log analytics. The underlying API that we are using to send out data into Log Analytics is documented here: https://docs.microsoft.com/en-us/azure/azure-monitor/logs/data-collector-api.

`Also, for the file in question, what is the operation we are applying using this logic app? This would help us define whether we should consider this as a playbook or a data connector or none.`
I am assuming you're referring to this file? https://raw.githubusercontent.com/Azure/Azure-Sentinel/07ac54eda970d0a464c2f0816edc749d08b9a877/Solutions/Tanium/Playbooks/Tanium-AddComment/azuredeploy.json
If so, this playbook is responsible for adding specific comments into an incident. Here is the photo of the logic-app designer, this should layout what it does. It simply takes in information, then posts the information into an incident inside of sentinel. 
![image](https://user-images.githubusercontent.com/11982273/178323008-89ee247e-74c4-47a5-a3fe-9b6182249aff.png)

Reply:
@devikamehra I added the connection as a resource in all the playbooks, and re-ran the packaging tool to create a 1.0.2.zip file. When I try to import the maintemplate.json into the webportal  I am seeing this error:
 
"Deployment template validation failed: 'The resource 'Microsoft.Web/connections/taniumsentinel' at line '262' and column '9' is defined multiple times in a template. Please see https://aka.ms/arm-template/#resources for usage details.'. (Code: InvalidTemplate)" 

Is there a way I can use a single connection and share it for all the playbooks?

Reply:
@ajcollins0 I tried deploying the mainTemplate.json file and it is working fine for me. But you should be using v2 of the packaging tool as we are moving all solutions from v1 to v2. Let me know if you need help in packaging.

As I see that the data lands into sentinel using push base mechanism, we would need a data connector to provide uniform experience to the sentinel user. This data connector will contain steps to configure data logging from Tanium Connector. Please use this as an example to develop the data connector - https://github.com/Azure/Azure-Sentinel/blob/master/Solutions/Sophos%20Cloud%20Optix/Data%20Connectors/Connector_REST_API_SophosCloudOptix.json

Kindly follow the guidance mentioned here to develop the data connector - https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/ReadMe.md

Place the developed JSON file in **Data Connector** folder (within Solution folder) and update the input file for packaging. Use the v2 of the packaging tool to develop the package. - https://github.com/Azure/Azure-Sentinel/tree/master/Tools/Create-Azure-Sentinel-Solution/V2

Reply:
Everything is working and getting the green light with the most recent push, and its using V2 of the packaging tool now. But when I attempt to import the mainTemplate.json, I am given this error and I don't know what this means. 


![image](https://user-images.githubusercontent.com/11982273/179143944-1ad2b784-b999-4d69-9439-fa70a2ab15bc.png)

Reply:
I should also note: Test-AzTemplate works on mainTemplate.json too, and the github actions have all passed.